File: android/app/src/main/kotlin/com/airclip/core/sync/ClipboardSyncEngine.kt
package com.airclip.core.sync

import com.airclip.core.clipboard.ClipContent
import com.airclip.core.clipboard.ClipKind
import com.airclip.core.clipboard.ClipboardOptions
import com.airclip.core.clipboard.ClipboardReadFailure
import com.airclip.core.clipboard.ClipboardReader
import com.airclip.core.clipboard.ClipboardWriter
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow

/** Where a publish attempt came from, so the UI can explain which backend actually works. */
enum class PublishSource { LISTENER, IME, ACCESSIBILITY, TILE, SHIZUKU, MANUAL, HISTORY }

sealed interface PublishResult {
    data class Sent(val content: ClipContent) : PublishResult

    /** Recognised as an echo of something we just received; deliberately dropped. */
    data object Suppressed : PublishResult

    data class Failed(val reason: ClipboardReadFailure) : PublishResult
}

/**
 * Wires a clipboard reader/writer pair to the [LoopGuard]. Platform-agnostic, like its Windows
 * counterpart `AirClip.Core.Sync.ClipboardSyncEngine` — Android just has several readers instead of
 * one message-loop monitor, so publishing is pull-based rather than event-based.
 */
class ClipboardSyncEngine(
    private val reader: ClipboardReader,
    private val writer: ClipboardWriter,
    val loopGuard: LoopGuard,
    private val options: () -> ClipboardOptions,
) {
    private val _published = MutableSharedFlow<ClipContent>(extraBufferCapacity = 16)
    private val _applied = MutableSharedFlow<ClipContent>(extraBufferCapacity = 16)
    private val _lastReadSource = MutableStateFlow<PublishSource?>(null)

    /** Local content that passed every filter and should go out to peers. */
    val published: SharedFlow<ClipContent> = _published.asSharedFlow()

    /** Remote content that was actually written to this device's clipboard. */
    val applied: SharedFlow<ClipContent> = _applied.asSharedFlow()

    /**
     * Which backend last read the clipboard successfully. The whole point of shipping four read
     * paths is that the user can see which one their ROM actually honours.
     */
    val lastReadSource: StateFlow<PublishSource?> = _lastReadSource.asStateFlow()

    suspend fun publishCurrent(source: PublishSource): PublishResult {
        val content = reader.read()
            ?: return PublishResult.Failed(reader.lastFailure ?: ClipboardReadFailure.EMPTY)
        _lastReadSource.value = source
        return publish(content, source)
    }

    suspend fun publish(content: ClipContent, source: PublishSource = PublishSource.MANUAL): PublishResult {
        withinLimits(content)?.let { return PublishResult.Failed(it) }

        if (!loopGuard.tryBeginPublish(content.hash)) {
            return PublishResult.Suppressed
        }

        _lastReadSource.value = source

        _published.emit(content)
        return PublishResult.Sent(content)
    }

    /**
     * Writes peer content to the local clipboard behind the loop guard. `false` means the write was
     * skipped (echo or over the limits) or the platform refused it.
     */
    suspend fun applyRemote(content: ClipContent): Boolean {
        if (withinLimits(content) != null) return false

        val scope = loopGuard.tryBeginApply(content.hash) ?: return false
        val written = try {
            writer.write(content)
        } finally {
            scope.close()
        }

        if (written) {
            _applied.emit(content)
        }
        return written
    }

    /**
     * Puts content back on the clipboard *without* republishing it — the history screen's "copy"
     * action. The hash is remembered so the resulting clipboard change is not bounced to peers.
     */
    suspend fun writeLocalOnly(content: ClipContent): Boolean {
        loopGuard.remember(content.hash)
        return writer.write(content)
    }

    /** `null` when the content may be synced, otherwise the reason it may not. */
    private fun withinLimits(content: ClipContent): ClipboardReadFailure? {
        val current = options()
        if (content.kind == ClipKind.IMAGE && !current.syncImages) {
            return ClipboardReadFailure.UNSUPPORTED_MIME
        }

        val limit = if (content.kind == ClipKind.TEXT) current.maxTextBytes else current.maxImageBytes
        return if (content.byteSize <= limit) null else ClipboardReadFailure.TOO_LARGE
    }
}

File: windows/src/AirClip.Core/Sync/ClipboardSyncEngine.cs
using AirClip.Core.Clipboard;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace AirClip.Core.Sync;

/// <summary>
/// Wires the platform clipboard monitor/writer to the loop guard. Platform-agnostic so the
/// filtering rules can be unit tested without touching a real clipboard.
/// </summary>
public sealed class ClipboardSyncEngine : IDisposable
{
    private readonly IClipboardMonitor _monitor;
    private readonly IClipboardWriter _writer;
    private readonly LoopGuard _loopGuard;
    private readonly ClipboardOptions _options;
    private readonly ILogger _logger;
    private bool _started;

    public ClipboardSyncEngine(
        IClipboardMonitor monitor,
        IClipboardWriter writer,
        LoopGuard loopGuard,
        ClipboardOptions? options = null,
        ILogger<ClipboardSyncEngine>? logger = null)
    {
        _monitor = monitor;
        _writer = writer;
        _loopGuard = loopGuard;
        _options = options ?? ClipboardOptions.Default;
        _logger = logger ?? NullLogger<ClipboardSyncEngine>.Instance;
    }

    /// <summary>Local clipboard content that passed every filter and should go out to peers.</summary>
    public event EventHandler<ClipboardChangedEventArgs>? LocalClipboardPublished;

    public void Start()
    {
        if (_started)
        {
            return;
        }

        _monitor.Changed += OnMonitorChanged;
        _monitor.Start();
        _started = true;
    }

    public void Stop()
    {
        if (!_started)
        {
            return;
        }

        _monitor.Changed -= OnMonitorChanged;
        _monitor.Stop();
        _started = false;
    }

    /// <summary>Manual "send clipboard now" path, used by tray menu and hotkeys.</summary>
    public bool PublishCurrent()
    {
        ClipboardContent? content = _monitor.ReadCurrent();
        return content is not null && TryPublish(content);
    }

    public async Task<bool> ApplyRemoteAsync(ClipboardContent content, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(content);

        if (!IsWithinLimits(content))
        {
            return false;
        }

        IDisposable? scope = _loopGuard.TryBeginApply(content.Hash);
        if (scope is null)
        {
            _logger.LogDebug("Skipped remote {Content}: already seen locally", content);
            return false;
        }

        using (scope)
        {
            await _writer.WriteAsync(content, cancellationToken).ConfigureAwait(false);
        }

        _logger.LogInformation("Applied remote {Content}", content);
        return true;
    }

    public void Dispose() => Stop();

    private void OnMonitorChanged(object? sender, ClipboardChangedEventArgs e) => TryPublish(e.Content);

    private bool TryPublish(ClipboardContent content)
    {
        if (!IsWithinLimits(content))
        {
            return false;
        }

        if (!_loopGuard.TryBeginPublish(content.Hash))
        {
            _logger.LogDebug("Suppressed local {Content}: echo of remote write", content);
            return false;
        }

        _logger.LogInformation("Publishing local {Content}", content);
        LocalClipboardPublished?.Invoke(this, new ClipboardChangedEventArgs(content));
        return true;
    }

    private bool IsWithinLimits(ClipboardContent content)
    {
        if (content.Kind == ClipboardContentKind.Image && !_options.SyncImages)
        {
            return false;
        }

        int limit = content.Kind == ClipboardContentKind.Text ? _options.MaxTextBytes : _options.MaxImageBytes;
        if (content.ByteSize <= limit)
        {
            return true;
        }

        _logger.LogWarning("Dropped {Content}: exceeds {Limit} bytes", content, limit);
        return false;
    }
}
